enhancements/catalog-data-transfer-reduction.md

```diff
@@ -0,0 +1,450 @@
+    - This is a candidate for follow-up work.
+1. Decide which architectures to mirror.
+
+## Proposal
+
+### User Stories [optional]
+
+<!-- TODO maybe -->
+
+### Implementation Details/Notes/Constraints [optional]
+
+This section delves into [declarative configs][declcfg-ep] (DC) heavily so it is recommended you read that EP first.
+
+Some terminology:
+- *Bundle*: a collection of manifests that define an operator at a particular state (ex. version).
+- *Replaces*: a [reference][replaces] to another bundle that can be upgraded from to the referent.
+- *Skips*: a [reference][skips] to another bundle that can be skipped when upgrading to the referent.
+- *Dependency*: a package or group-version-kind (GVK) requirement by a bundle.
+  - There are three ways to specify dependencies*: by full package, by package version range, and by GVK.
+  The first way is effectively the same as the second, with version range being the entire set of published versions.
+- *Upgrade graph*: the DAG of replaces from one bundle to another, skipping bundles referenced in skips.
+- *Replaces chain*: upgrade graph, ignoring skips.
+- *Channel*: a named upgrade graph.
+- *Channel head*: a bundle in a channel that is not referenced by any replaces or skips.
+- *Package*: a collection of channels.
+- *Index*: a collection of packages, each of which may or may not contain all published bundles for that package.
+  - Typically built into an image that can serve packages via gRPC when run to OLM via a CatalogSource
+- *Catalog*: one or more indices serving content.
+
+Catalog invariants:
+- Bundles, channels, and packages have unique names within their parent type
+(bundle->channel->package->catalog).
+- A package name + bundle name is a unique identifier within a catalog.
+- A CSV name is unique to a set of catalogs. That is, a catalog is invalid if
+two packages contain CSVs with the same name.
+    - This may be removed in favor of package + bundle name only.
+- A channel is invalid if it does not have exactly one head.
+    - This may be removed at some point, but can be managed by iterating through each head.
+
+It is also worth noting that an operator will fail to install at runtime if a dependency is not met.
+This is not an enforced catalog invariant because a dependency may be satisfied by any available catalog,
+a fact that cannot be known until runtime.
+
+Given the above [goals](#goals), a process that calculates all bundles published between two catalog states is needed.
+The old state is a full catalog or set of packages, channels, and versions (“catalog objects”);
+the new state is a full catalog containing the old state plus newly published catalog objects.
+
+A new `opm diff` command will be introduced, with accompanying libraries, to perform this calculation.
+The output will be a declarative config. There are two modes to `opm diff`:
+- *FromHeads*: diff from input channel heads, and exclude heads from the resulting diff.
```

Comment:
Should this say "exclude the head from the old state"?

Reply:
Yes, good catch.